infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -143,75 +154,46 @@
                                             @Nonnull CorfuReplicationClusterManagerAdapter clusterManagerAdapter,
                                             @Nonnull CompletableFuture<CorfuInterClusterReplicationServerNode> serverCallback) {
+        this.executorService = Executors.newSingleThreadExecutor(new
+                ThreadFactoryBuilder().setNameFormat("discovery-service-executor").build());
         this.clusterManagerAdapter = clusterManagerAdapter;
         this.logReplicationNodeId = serverContext.getNodeId();
         this.serverContext = serverContext;
         this.localEndpoint = serverContext.getLocalEndpoint();
         this.serverCallback = serverCallback;
+        this.shutdown = false;
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.executorService.submit(this::run);
         this.isLeader = new AtomicBoolean();
     }
 
-    public void run() {
-        try {
-            startDiscovery();
-
-            while (shouldRun) {
-                try {
-                    DiscoveryServiceEvent event = eventQueue.take();
-                    processEvent(event);
-                } catch (Exception e) {
-                    log.error("Caught an exception. Stop discovery service.", e);
-                    shouldRun = false;
-                    stopLogReplication();
-                    if (e instanceof InterruptedException) {
-                        Thread.interrupted();
-                    }
-                }
-            }
-        } catch (Exception e) {
-            log.error("Unhandled exception caught during log replication service discovery. Retry,", e);
-        } finally {
-            if (runtime != null) {
-                runtime.shutdown();
-            }
-        }
-    }
-
     /**
-     * Process discovery event
+     * The executor thread will keep running at background and process events in eventQueue
      */
-    public void processEvent(DiscoveryServiceEvent event) {
-        switch (event.type) {
-            case ACQUIRE_LOCK:
-                processLockAcquire();
-                break;
-
-            case RELEASE_LOCK:
-                processLockRelease();
-                break;
-
-            case DISCOVERED_TOPOLOGY:
-                processTopologyChangeNotification(event);
-                break;
+    public void run() {
 
-            case UPGRADE:
-                processUpgrade(event);
-                break;
+        // The DISCOVERY_INIT_TOPOLOGY should be the first event to be processed for both ACTIVE and STANDBY sites.
+        this.eventQueue.add(new DiscoveryServiceEvent(DISCOVER_INIT_TOPOLOGY));
 
-            default:
-                log.error("Invalid event type {}", event.type);
+        while (!shutdown) {
+            try {
+                DiscoveryServiceEvent event = eventQueue.take();
+                processEvent(event);
+            } catch (Exception | LogReplicationDiscoveryServiceException e) {
```

Comment:
We don't have LogReplicationDiscoveryServiceException anymore. It is good to do some cleanup.